### ISSUE TITLE
Multiple subscriptions question

### ISSUE THREAD

**Issue by pantonis** [stomp-js/ng2-stompjs#2]
Hi,

Using one instance of a StompService can I subscribe to multiple queues?

**Comment by kum-deepak** [comment]
Yes, you can subscribe any number of queues. Just make multiple calls to subscribe.

**Comment by pantonis** [comment]
And each subscription must be assigned to a different observable?

**Comment by kum-deepak** [comment]
When you make a call to subscribe, a new Observable is returned, which will provide messages for that queue.

**Comment by pantonis** [comment]
Forgive me I am not web dev expert.
````
 this.messages = this._stompService.subscribe('/topic/ng-demo-sub');
 this.messages2 = this._stompService.subscribe('/topic/ng-demo-sub2');
````
or 
````
 this.messages = this._stompService.subscribe('/topic/ng-demo-sub');
 this.messages = this._stompService.subscribe('/topic/ng-demo-sub2');
````

**Comment by kum-deepak** [comment]
The first one:

this.messages = this._stompService.subscribe('/topic/ng-demo-sub');
 this.messages2 = this._stompService.subscribe('/topic/ng-demo-sub2');

**Comment by pantonis** [comment]
and to unsubscribe 
```
this.subscription.unsubscribe();
```

is this enough for both subscriptions?

**Comment by kum-deepak** [comment]
No, the code will look something like following (I haven't executed the code, just typed here):

```typescript
this.messages = this._stompService.subscribe('/topic/ng-demo-sub');
this.messages2 = this._stompService.subscribe('/topic/ng-demo-sub2');

// use messages from /topic/ng-demo-sub by subscribing to the Observable
this.messages.subscribe((msg: Message) => {
   // use it
});

// use messages from /topic/ng-demo-sub2 by subscribing to the Observable
this.messages2.subscribe((msg: Message) => {
   // use it
});

// Unsubscribe from /topic/ng-demo-sub
this.messages.unsubscribe();
this.messages = null;

// Unsubscribe from /topic/ng-demo-sub2
this.messages2.unsubscribe();
this.messages2 = null;
```

**Comment by kum-deepak** [comment]
In summary both the subscriptions are independent of each other and do not affect each other.

**Comment by pantonis** [comment]
ok thanks a lot

**Comment by kum-deepak** [comment]
@marcelormourao, please create a demo project demonstrating the issue. You can fork one of the demo projects (links in the README). Thanks!

Meanwhile I am also taking a look.

**Comment by fernandoGN** [comment]
I has a similiar problem =(

**Comment by kum-deepak** [comment]
I have tested the following sequence (I have added these and few more cases in `src/specs/app/services/stomp-queue.operations.spec.ts`):

```typescript
      queSubscription1 = stompService.subscribe(queueName1)
        .map((message) => message.body)
        .subscribe(spyHandler1);
      queSubscription2 = stompService.subscribe(queueName2)
        .map((message) => message.body)
        .subscribe(spyHandler2);

      stompService.publish(queueName1, 'Message 01-01');
      stompService.publish(queueName2, 'Message 02-01');
     
      queSubscription1.unsubscribe();

      stompService.publish(queueName1, 'Message 01-02');
      stompService.publish(queueName2, 'Message 02-02');
```
Please see the log output corresponding to the above call. Lines starting with **** are my annotations.

```text
LOG: Fri Apr 13 2018 14:56:03 GMT+0530 (IST), 'Opening Web Socket...'
LOG: Fri Apr 13 2018 14:56:03 GMT+0530 (IST), 'Connecting...'
LOG: Fri Apr 13 2018 14:56:03 GMT+0530 (IST), 'Web Socket Opened...'
LOG: Fri Apr 13 2018 14:56:03 GMT+0530 (IST), '>>> CONNECT
login:guest
passcode:guest
accept-version:1.2,1.1,1.0
heart-beat:0,0

**Comment by kum-deepak** [comment]
Please note that I have updated my schematic code at https://github.com/stomp-js/ng2-stompjs/issues/2#issuecomment-293604982

**Comment by deli6z** [comment]
@kum-deepak I am having the similar issue. I reported it here https://github.com/stomp-js/ng2-stompjs/issues/73. Basically it does not work if I specify a subscription callback via an inline function then I am having that issue. I did not try anonymous function like you had in your example - that might work but I still would like to be able to set inline function as a callback to the topic Observable subscribe. Thanks

**Comment by s1vert** [comment]
If you pass stomp headers to `subscribe` method like `stompService.subscribe(destination, headers)`, make sure you create new headers with each call to `subscribe`. Headers are modified inside and that's causing problems with subscriptions.

**Comment by kum-deepak** [comment]
@s1vert, seems bit peculiar behavior. Please fork any of the samples and modify it to show the issue. I will like to have a look at the issue.

**Comment by kum-deepak** [comment]
@s1vert I think I undertsand the issue now. It will need to be fixed in underlying library as well as in this one. It will be done in next major release.

**Comment by sahyun1** [comment]
would this be part of 7.0.0?

**Comment by kum-deepak** [comment]
@sahyun1 this thread has two different issues mentioned. The original question - multiple subscriptions - that has always worked.

There was one more issue that some functions were altering the header argument. It has been fixed in underlying library https://github.com/stomp-js/stompjs/issues/11. It will be part of v7.

If you mean any other issue, please create a new one 😄

**Comment by ramakanthreddyk** [comment]
> No, the code will look something like following (I haven't executed the code, just typed here):
> 
> ```ts
> // use messages from /topic/ng-demo-sub by subscribing to the Observable
> this.messages = this._stompService
>   .subscribe('/topic/ng-demo-sub')
>   .subscribe((msg: Message) => {
>     // use it
>   });
> 
> // use messages from /topic/ng-demo-sub2 by subscribing to the Observable
> this.messages2 = this._stompService
>   .subscribe('/topic/ng-demo-sub2')
>   .subscribe((msg: Message) => {
>     // use it
>   });
> 
> // Unsubscribe from /topic/ng-demo-sub
> this.messages.unsubscribe();
> this.messages = null;
> 
> // Unsubscribe from /topic/ng-demo-sub2
> this.messages2.unsubscribe();
> this.messages2 = null;
> ```
> 
> The above has been updated.

Cant we do it dynamically ?

**Comment by kum-deepak** [comment]
Could not understand your query. Please post some code sample.

**Comment by dJani97** [comment]
> And each subscription must be assigned to a different observable?

If the number of new variables gets out of hand, you may use an array to collect all the subscriptions and use a loop to unsubscribe in the end. This also guarantees that you don't forget about any subscriptions.

This looks like a good place to use a pattern like [subsink](https://www.npmjs.com/package/subsink). You just throw a bunch of subscriptions at it, then call unsubscribe once. The code behind this library is dead simple, if it doesn't suit you, it's pretty easy to make your own implementation.
